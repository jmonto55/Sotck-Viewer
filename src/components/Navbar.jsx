import { Link } from 'react-router-dom';
import { useState, useEffect, useRef } from 'react';

const Navbar = () => {
  const [markets, setMarkets] = useState(false);
  const [open, setOpen] = useState(false);
  const ref1 = useRef();
  const ref2 = useRef();

  const toggleMarkets = () => {
    setMarkets(!markets);
  };

  const toggleOpen = () => {
    setOpen(!open);
  };

  useEffect(() => {
    const handler = (event) => {
      if (markets && ref1.current && !ref1.current.contains(event.target)) {
        setMarkets(false);
      }
      if (open && ref2.current && !ref2.current.contains(event.target)) {
        setOpen(false);
      }
    };
    document.addEventListener('mousedown', handler);
  }, [open, markets]);

  return (
    <div className="flex justify-center">
      <nav className="z-20 bg-neutral-900 fixed flex w-full max-w-screen-2xl justify-between px-8 md:px-24 py-5 m-auto text-neutral-100">
        <Link className="text-xl md:text-2xl" to="/">
          Quantum
          <span className="font-bold">StocksðŸ’±</span>
        </Link>
        <ul
          ref={ref2}
          className={
            `${open ? '-translate-y-2.5' : 'right-0 -translate-y-64'} bg-neutral-900 md:text-xl
            ease-in-out duration-700 ${markets ? '' : 'rounded-b-lg'} fixed right-0 top-20
            w-32 flex flex-col justify-between items-center md:static md:translate-y-0
            md:flex-row md:w-48 md:justify-between md:items-center md:right-base shadow-lg shadow-black md:shadow-transparent`
          }
        >
          <li className="mb-4 md:mb-0">
            <Link onClick={toggleOpen} className="hover:text-gray-400" to="/">Home</Link>
          </li>
          <li>
            <button onClick={toggleMarkets} type="submit" className="mb-4 md:mb-0 flex cursor-pointer hover:text-gray-400">
              Markets
              <span ref={ref1} className="text-base material-symbols-outlined cursor-pointer motion-safe:animate-bounce-slow mt-1">
                expand_more
              </span>
            </button>
            <ul
              className={`flex flex-col fixed shadow-5xl p-2 w-32 rounded-b-lg right-0 bg-neutral-800
            ${markets ? 'translate-x-0 -translate-y-2' : 'translate-x-full opacity-0'} ease-in-out duration-700 md:mt-2
            backdrop-filter backdrop-blur-sm shadow-lg shadow-black`}
            >
              <Link onClick={toggleOpen} className="mb-3 border-b hover:text-gray-400" to="/market">NYSE</Link>
              <Link onClick={toggleOpen} className="mb-3 border-b hover:text-gray-400" to="/market">NASDAQ</Link>
              <Link onClick={toggleOpen} className="mb-3 border-b hover:text-gray-400" to="/market">EURONEXT</Link>
              <Link onClick={toggleOpen} className="mb-3 border-b hover:text-gray-400" to="/market">AMEX</Link>
              <Link onClick={toggleOpen} className="mb-1 hover:text-gray-400" to="/market">TSX</Link>
            </ul>
          </li>
        </ul>
        {open && (
        <button onClick={toggleOpen} type="button" className="z-10 md:hidden">
          <span className="material-symbols-outlined">
            close
          </span>
        </button>
        )}
        {!open && (
        <button onClick={toggleOpen} type="button" className="z-10 md:hidden">
          <span className="material-symbols-outlined">
            menu
          </span>
        </button>
        )}
      </nav>
    </div>
  );
};

export default Navbar;
