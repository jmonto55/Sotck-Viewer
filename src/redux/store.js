import { configureStore } from '@reduxjs/toolkit';
import marketsReducer from './markets/marketsSlice';

const store = configureStore({
  reducer: {
    markets: marketsReducer,
  },
});

export default store;
