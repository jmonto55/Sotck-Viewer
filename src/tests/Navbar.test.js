import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import Navbar from '../components/Navbar';
import store from '../redux/store';
import '@testing-library/jest-dom';

describe('Test suite for Navbar component', () => {
  it('Should render a Snapshot of Navbar component', () => {
    const navbar = render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>,
    );
    expect(navbar).toMatchSnapshot();
  });
  it('Should test if the logo is being rendered', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>,
    );

    expect(screen.getByText('Quantum')).toBeInTheDocument();
  });

  it('Should test if the Navbar links are being rendered', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>,
    );

    const etfLink = screen.getByText(/etf/i);
    expect(etfLink).toBeInTheDocument();

    const nyseLink = screen.getByText(/nyse/i);
    expect(nyseLink).toBeInTheDocument();

    const euronextLink = screen.getByText(/euronext/i);
    expect(euronextLink).toBeInTheDocument();

    const amexLink = screen.getByText(/amex/i);
    expect(amexLink).toBeInTheDocument();

    const tsxLink = screen.getByText(/tsx/i);
    expect(tsxLink).toBeInTheDocument();

    const nasdaqLink = screen.getByText(/nasdaq/i);
    expect(nasdaqLink).toBeInTheDocument();
  });

  it('Should dispatch the selectMarket action when a link is clicked', () => {
    const selectMarketMock = jest.fn();
    store.dispatch = selectMarketMock;

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>,
    );

    const etfLink = screen.getByText(/etf/i);
    fireEvent.click(etfLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'ETF', type: 'markets/selectMarket' });

    const nyseLink = screen.getByText(/nyse/i);
    fireEvent.click(nyseLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'NYSE', type: 'markets/selectMarket' });

    const euronextLink = screen.getByText(/euronext/i);
    fireEvent.click(euronextLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'EURONEXT', type: 'markets/selectMarket' });

    const amexLink = screen.getByText(/amex/i);
    fireEvent.click(amexLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'AMEX', type: 'markets/selectMarket' });

    const tsxLink = screen.getByText(/tsx/i);
    fireEvent.click(tsxLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'TSX', type: 'markets/selectMarket' });

    const nasdaqLink = screen.getByText(/nasdaq/i);
    fireEvent.click(nasdaqLink);
    expect(selectMarketMock).toHaveBeenCalledWith({ payload: 'NASDAQ', type: 'markets/selectMarket' });
  });
});
