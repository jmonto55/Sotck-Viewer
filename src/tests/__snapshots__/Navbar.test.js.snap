// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Navbar component Should render a Snapshot of Navbar component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex justify-center"
      >
        <nav
          class="z-30 bg-neutral-900 fixed flex w-full max-w-screen-2xl justify-between px-8 md:px-24 py-7 m-auto text-neutral-100"
        >
          <a
            class="text-xl md:text-2xl"
            href="/"
          >
            Quantum
            <span
              class="font-bold"
            >
              StocksðŸ’±
            </span>
          </a>
          <ul
            class="right-0 -translate-y-64 bg-neutral-900 md:text-xl
            ease-in-out duration-700 rounded-b-lg fixed right-0 top-20
            w-32 flex flex-col justify-between items-center md:static md:translate-y-0
            md:flex-row md:w-48 md:justify-between md:items-center md:right-base shadow-lg shadow-black md:shadow-transparent"
          >
            <li
              class="mb-4 md:mb-0"
            >
              <a
                class="hover:text-gray-400"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <button
                class="mb-4 md:mb-0 flex cursor-pointer hover:text-gray-400"
                type="submit"
              >
                Markets
                <span
                  class="text-base material-symbols-outlined cursor-pointer motion-safe:animate-bounce-slow mt-1"
                >
                  expand_more
                </span>
              </button>
              <ul
                class="flex flex-col fixed shadow-5xl p-2 w-32 rounded-b-lg right-0 bg-neutral-800
            translate-x-full opacity-0 ease-in-out duration-700 md:mt-2
            backdrop-filter backdrop-blur-sm shadow-lg shadow-black"
              >
                <a
                  class="mb-3 border-b hover:text-gray-400"
                  href="/market"
                >
                  ETF
                </a>
                <a
                  class="mb-3 border-b hover:text-gray-400"
                  href="/market"
                >
                  NYSE
                </a>
                <a
                  class="mb-3 border-b hover:text-gray-400"
                  href="/market"
                >
                  EURONEXT
                </a>
                <a
                  class="mb-3 border-b hover:text-gray-400"
                  href="/market"
                >
                  AMEX
                </a>
                <a
                  class="mb-3 border-b hover:text-gray-400"
                  href="/market"
                >
                  TSX
                </a>
                <a
                  class="mb-1 hover:text-gray-400"
                  href="/market"
                >
                  NASDAQ
                </a>
              </ul>
            </li>
          </ul>
          <button
            class="z-10 md:hidden"
            type="button"
          >
            <span
              class="material-symbols-outlined"
            >
              menu
            </span>
          </button>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex justify-center"
    >
      <nav
        class="z-30 bg-neutral-900 fixed flex w-full max-w-screen-2xl justify-between px-8 md:px-24 py-7 m-auto text-neutral-100"
      >
        <a
          class="text-xl md:text-2xl"
          href="/"
        >
          Quantum
          <span
            class="font-bold"
          >
            StocksðŸ’±
          </span>
        </a>
        <ul
          class="right-0 -translate-y-64 bg-neutral-900 md:text-xl
            ease-in-out duration-700 rounded-b-lg fixed right-0 top-20
            w-32 flex flex-col justify-between items-center md:static md:translate-y-0
            md:flex-row md:w-48 md:justify-between md:items-center md:right-base shadow-lg shadow-black md:shadow-transparent"
        >
          <li
            class="mb-4 md:mb-0"
          >
            <a
              class="hover:text-gray-400"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <button
              class="mb-4 md:mb-0 flex cursor-pointer hover:text-gray-400"
              type="submit"
            >
              Markets
              <span
                class="text-base material-symbols-outlined cursor-pointer motion-safe:animate-bounce-slow mt-1"
              >
                expand_more
              </span>
            </button>
            <ul
              class="flex flex-col fixed shadow-5xl p-2 w-32 rounded-b-lg right-0 bg-neutral-800
            translate-x-full opacity-0 ease-in-out duration-700 md:mt-2
            backdrop-filter backdrop-blur-sm shadow-lg shadow-black"
            >
              <a
                class="mb-3 border-b hover:text-gray-400"
                href="/market"
              >
                ETF
              </a>
              <a
                class="mb-3 border-b hover:text-gray-400"
                href="/market"
              >
                NYSE
              </a>
              <a
                class="mb-3 border-b hover:text-gray-400"
                href="/market"
              >
                EURONEXT
              </a>
              <a
                class="mb-3 border-b hover:text-gray-400"
                href="/market"
              >
                AMEX
              </a>
              <a
                class="mb-3 border-b hover:text-gray-400"
                href="/market"
              >
                TSX
              </a>
              <a
                class="mb-1 hover:text-gray-400"
                href="/market"
              >
                NASDAQ
              </a>
            </ul>
          </li>
        </ul>
        <button
          class="z-10 md:hidden"
          type="button"
        >
          <span
            class="material-symbols-outlined"
          >
            menu
          </span>
        </button>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
